// @editorjs/dom/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.prepend = exports.offset = exports.make = exports.isLineBreakTag = exports.isSingleTag = exports.isNodeEmpty = exports.isLeaf = exports.isHTMLString = exports.isFragment = exports.isEmpty = exports.isElement = exports.isContentEditable = exports.isCollapsedWhitespaces = exports.findAllInputs = exports.isNativeInput = exports.allInputsSelector = exports.getDeepestNode = exports.getDeepestBlockElements = exports.getContentLength = exports.fragmentToString = exports.containsOnlyInlineElements = exports.canSetCaret = exports.calculateBaseline = exports.blockElements = exports.append = void 0;
var allInputsSelector_1 = require("./allInputsSelector");
Object.defineProperty(exports, "allInputsSelector", { enumerable: true, get: function () { return allInputsSelector_1.allInputsSelector; } });
var isNativeInput_1 = require("./isNativeInput");
Object.defineProperty(exports, "isNativeInput", { enumerable: true, get: function () { return isNativeInput_1.isNativeInput; } });
var append_1 = require("./append");
Object.defineProperty(exports, "append", { enumerable: true, get: function () { return append_1.append; } });
var blockElements_1 = require("./blockElements");
Object.defineProperty(exports, "blockElements", { enumerable: true, get: function () { return blockElements_1.blockElements; } });
var calculateBaseline_1 = require("./calculateBaseline");
Object.defineProperty(exports, "calculateBaseline", { enumerable: true, get: function () { return calculateBaseline_1.calculateBaseline; } });
var canSetCaret_1 = require("./canSetCaret");
Object.defineProperty(exports, "canSetCaret", { enumerable: true, get: function () { return canSetCaret_1.canSetCaret; } });
var containsOnlyInlineElements_1 = require("./containsOnlyInlineElements");
Object.defineProperty(exports, "containsOnlyInlineElements", { enumerable: true, get: function () { return containsOnlyInlineElements_1.containsOnlyInlineElements; } });
var fragmentToString_1 = require("./fragmentToString");
Object.defineProperty(exports, "fragmentToString", { enumerable: true, get: function () { return fragmentToString_1.fragmentToString; } });
var getContentLength_1 = require("./getContentLength");
Object.defineProperty(exports, "getContentLength", { enumerable: true, get: function () { return getContentLength_1.getContentLength; } });
var getDeepestBlockElements_1 = require("./getDeepestBlockElements");
Object.defineProperty(exports, "getDeepestBlockElements", { enumerable: true, get: function () { return getDeepestBlockElements_1.getDeepestBlockElements; } });
var getDeepestNode_1 = require("./getDeepestNode");
Object.defineProperty(exports, "getDeepestNode", { enumerable: true, get: function () { return getDeepestNode_1.getDeepestNode; } });
var findAllInputs_1 = require("./findAllInputs");
Object.defineProperty(exports, "findAllInputs", { enumerable: true, get: function () { return findAllInputs_1.findAllInputs; } });
var isCollapsedWhitespaces_1 = require("./isCollapsedWhitespaces");
Object.defineProperty(exports, "isCollapsedWhitespaces", { enumerable: true, get: function () { return isCollapsedWhitespaces_1.isCollapsedWhitespaces; } });
var isContentEditable_1 = require("./isContentEditable");
Object.defineProperty(exports, "isContentEditable", { enumerable: true, get: function () { return isContentEditable_1.isContentEditable; } });
var isElement_1 = require("./isElement");
Object.defineProperty(exports, "isElement", { enumerable: true, get: function () { return isElement_1.isElement; } });
var isEmpty_1 = require("./isEmpty");
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return isEmpty_1.isEmpty; } });
var isFragment_1 = require("./isFragment");
Object.defineProperty(exports, "isFragment", { enumerable: true, get: function () { return isFragment_1.isFragment; } });
var isHtmlString_1 = require("./isHtmlString");
Object.defineProperty(exports, "isHTMLString", { enumerable: true, get: function () { return isHtmlString_1.isHTMLString; } });
var isLeaf_1 = require("./isLeaf");
Object.defineProperty(exports, "isLeaf", { enumerable: true, get: function () { return isLeaf_1.isLeaf; } });
var isNodeEmpty_1 = require("./isNodeEmpty");
Object.defineProperty(exports, "isNodeEmpty", { enumerable: true, get: function () { return isNodeEmpty_1.isNodeEmpty; } });
var isLineBreakTag_1 = require("./isLineBreakTag");
Object.defineProperty(exports, "isLineBreakTag", { enumerable: true, get: function () { return isLineBreakTag_1.isLineBreakTag; } });
var isSingleTag_1 = require("./isSingleTag");
Object.defineProperty(exports, "isSingleTag", { enumerable: true, get: function () { return isSingleTag_1.isSingleTag; } });
var make_1 = require("./make");
Object.defineProperty(exports, "make", { enumerable: true, get: function () { return make_1.make; } });
var offset_1 = require("./offset");
Object.defineProperty(exports, "offset", { enumerable: true, get: function () { return offset_1.offset; } });
var prepend_1 = require("./prepend");
Object.defineProperty(exports, "prepend", { enumerable: true, get: function () { return prepend_1.prepend; } });
